File: backend/domain/goal/learning_goal.py
from __future__ import annotations

from datetime import UTC, datetime
from uuid import uuid4

from pydantic import BaseModel, Field

from backend.domain.goal.goal_status import GoalStatus


class LearningGoal(BaseModel):
    """
    Aggregate Root representing a learner's goal.

    A goal drives workflow planning and progress tracking.
    """

    id: str = Field(
        default_factory=lambda: str(uuid4()),
    )

    student_id: str

    knowledge_node_id: str

    target_level: str

    status: GoalStatus = GoalStatus.NOT_STARTED

    created_at: datetime = Field(
        default_factory=lambda: datetime.now(
            UTC,
        ),
    )

    completed_at: datetime | None = None

    def start(
        self,
    ) -> None:

        self.status = GoalStatus.IN_PROGRESS

    def complete(
        self,
    ) -> None:

        self.status = GoalStatus.COMPLETED
        self.completed_at = datetime.now(
            UTC,
        )

    def pause(
        self,
    ) -> None:

        self.status = GoalStatus.PAUSED

    def cancel(
        self,
    ) -> None:

        self.status = GoalStatus.CANCELLED

File: frontend/lib/domain/models/goal/learning_goal.dart
import 'goal_status.dart';

class LearningGoal {
  final String id;
  final String studentId;
  final String knowledgeNodeId;
  final String targetLevel;
  final GoalStatus status;
  final DateTime createdAt;
  final DateTime? completedAt;

  const LearningGoal({
    required this.id,
    required this.studentId,
    required this.knowledgeNodeId,
    required this.targetLevel,
    this.status = GoalStatus.notStarted,
    required this.createdAt,
    this.completedAt,
  });
}
